src/persist/src/operators/stream.rs

```diff
@@ -534,29 +558,66 @@ where
                                         &operator_name,
                                         pending_future.cap.time(),
                                     );
+
+                                    // Drop any previous seal futures that are superseded by this
+                                    // successful seal. This will clean out seals that might have
+                                    // errored but are still holding back the frontier by holding
+                                    // on to a capability.
+                                    let seal_ts = pending_future.cap.time();
+                                    pending_futures.retain(|seal_future| match seal_future {
+                                        PendingSealFuture::PrimarySeal(seal_future) => {
+
+                                            if seal_future.cap.time() <= seal_ts {
+                                                log::trace!("Dropping seal (primary) {} because it is superseded by {}.", seal_future.cap.time(), seal_ts);
+                                                false
+                                            } else {
+                                                log::trace!("Retaining seal (primary) {} because it is NOT superseded by {}.", seal_future.cap.time(), seal_ts);
+                                                true
+                                            }
+                                        }
+                                        // Primary seals don't supersede condition seals.
```

Comment:
why not? it feels like they could / should?

Reply:
Because we're sealing two different collections. Successfully sealing the _condition_ collection to time `t` doesn't mean that we sealed the _primary_ collection to time `t`, so we shouldn't release data from the primary input yet.

Reply:
Ah, but also: we will only ever have a successful seal for primary on time `t` when we have already sealed condition up to `t`. So this comment might be moot.